### ISSUE TITLE
Garbage collection impacting performance when using multiple isolates and the camera.

### ISSUE THREAD

**Issue by martin-labanic** [flutter/flutter#48360]
I am attempting to process image data on one isolate provided by a camera preview running in the main isolate. While I expected garbage collection to have some impact it appears to be causing my performance to plummet, and this is before I attempt to run it through a third party c++ library via the dart ffi.

## Details

To reproduce the problem just run the provided sample project for a minute and look at the output.

This is a stripped down sample project using the camera package and the dart ffi that reproduces the issue at a smaller scale. When I have an image processing library in the c++ it averages at 160ms but I've seen it hit as high as 600ms on the 5th gen iPad to process 1 frame on the image processing isolate. The third party library I'm using typically takes ~30ms to process one image on this device so its a significant discrepancy.
https://github.com/martin-labanic/camera_preview_ffi_image_processing

## Update (2020/01/17):
Android is performing as well as our current app native app, iOS still has problems. After looking at the newest traces with the latest master channel changes, it seems like the problem has to do with `MessageLoop::FlushTasks`. I'm not certain on whether or not that is part of garbage collection.

iOS timeline traces:
Old: 
5th gen iPad: [example_project_dart_timeline_2020_01_08.json.zip](https://github.com/flutter/flutter/files/4036309/example_project_dart_timeline_2020_01_08.json.zip)

New:
5th gen iPad: [example_project_dart_timeline_2020_01_17_5th_gen_ipad.json.zip](https://github.com/flutter/flutter/files/4078232/example_project_dart_timeline_2020_01_17_5th_gen_ipad.json.zip)

7th gen iPad: [example_project_dart_timeline_2020_01_17_7th_gen_ipad.json.zip](https://github.com/flutter/flutter/files/4078538/example_project_dart_timeline_2020_01_17_7th_gen_ipad.json.zip)

Target Platform: iOS & Android
Target OS version/browser: ios >= 11, android >= API 23
Devices: 5th gen ipad, sm-t290

## Logs

devtools profiling build output from a 5th gen iPad:
```
.
.
.
10:17:27.289	stdout	flutter: -------------------------- startImageStream: processed camera data in 9ms --------------------------
10:17:27.310	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047309
10:17:27.314	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047314
10:17:27.315	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.316	stdout	flutter: ZZZZZ processImage: ffi call: 1ms
10:17:27.316	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047316
10:17:27.317	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047317
10:17:27.317	stdout	flutter: -------------------------- startImageStream: processed camera data in 8ms --------------------------
10:17:27.351	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047351
10:17:27.360	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047360
10:17:27.363	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 3ms
10:17:27.364	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.364	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047364
10:17:27.364	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047364
10:17:27.365	stdout	flutter: -------------------------- startImageStream: processed camera data in 13ms --------------------------
10:17:27.385	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047385
10:17:27.390	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047390
10:17:27.391	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.391	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.391	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047391
10:17:27.392	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047392
10:17:27.392	stdout	flutter: -------------------------- startImageStream: processed camera data in 7ms --------------------------
10:17:27.418	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047418
10:17:27.426	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047426
10:17:27.427	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.427	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.430	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047428
10:17:27.430	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047430
10:17:27.430	stdout	flutter: -------------------------- startImageStream: processed camera data in 12ms --------------------------
10:17:27.457	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047457
10:17:27.470	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047470
10:17:27.471	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 0ms
10:17:27.471	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.471	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047471
10:17:27.472	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047472
10:17:27.472	stdout	flutter: -------------------------- startImageStream: processed camera data in 15ms --------------------------
10:17:27.499	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047499
10:17:27.501	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047501
10:17:27.502	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.502	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.502	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047502
10:17:27.534	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047533
10:17:27.534	stdout	flutter: -------------------------- startImageStream: processed camera data in 35ms --------------------------
10:17:27.536	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047536
10:17:27.546	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047546
10:17:27.547	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.547	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.548	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047548
10:17:27.549	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047549
10:17:27.549	stdout	flutter: -------------------------- startImageStream: processed camera data in 13ms --------------------------
10:17:27.671	gc	main • finalize collection in 3002 ms • 606.6 MB used of 618.8 MB
10:17:27.673	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047673
10:17:27.677	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047677
10:17:27.678	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.678	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.678	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047678
10:17:27.679	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047679
10:17:27.680	stdout	flutter: -------------------------- startImageStream: processed camera data in 6ms --------------------------
10:17:27.699	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047699
10:17:27.701	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047701
10:17:27.702	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.703	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.703	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047703
10:17:27.704	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047703
10:17:27.704	stdout	flutter: -------------------------- startImageStream: processed camera data in 5ms --------------------------
10:17:27.732	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047732
10:17:27.734	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047734
10:17:27.735	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.735	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.736	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047736
10:17:27.736	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047736
10:17:27.737	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:27.766	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047765
10:17:27.768	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047768
10:17:27.768	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 0ms
10:17:27.769	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.769	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047769
10:17:27.770	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047770
10:17:27.770	stdout	flutter: -------------------------- startImageStream: processed camera data in 5ms --------------------------
10:17:27.799	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047799
10:17:27.801	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047801
10:17:27.802	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.802	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.802	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047802
10:17:27.803	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047803
10:17:27.803	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:27.832	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047832
10:17:27.834	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047834
10:17:27.835	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.835	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.836	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047835
10:17:27.836	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047836
10:17:27.837	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:27.866	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047865
10:17:27.871	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047871
10:17:27.872	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.872	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.873	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047873
10:17:27.873	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047873
10:17:27.873	stdout	flutter: -------------------------- startImageStream: processed camera data in 8ms --------------------------
10:17:27.899	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047899
10:17:27.902	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047901
10:17:27.902	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.903	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.903	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047903
10:17:27.903	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047903
10:17:27.903	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:27.933	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047932
10:17:27.935	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047934
10:17:27.935	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.935	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.936	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047936
10:17:27.936	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047936
10:17:27.937	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:27.966	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047966
10:17:27.968	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285047967
10:17:27.968	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:27.968	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:27.969	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285047969
10:17:27.969	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285047969
10:17:27.970	stdout	flutter: -------------------------- startImageStream: processed camera data in 3ms --------------------------
10:17:27.999	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285047999
10:17:28.001	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285048001
10:17:28.002	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:28.002	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:28.002	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285048002
10:17:28.003	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285048003
10:17:28.003	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:28.032	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285048032
10:17:28.034	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285048034
10:17:28.035	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:28.035	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:28.036	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285048036
10:17:28.038	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285048038
10:17:28.038	stdout	flutter: -------------------------- startImageStream: processed camera data in 6ms --------------------------
10:17:28.066	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285048066
10:17:28.068	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285048068
10:17:28.069	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:28.069	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:28.069	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285048069
10:17:28.070	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285048070
10:17:28.070	stdout	flutter: -------------------------- startImageStream: processed camera data in 4ms --------------------------
10:17:28.099	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285048099
10:17:28.101	gc	main • low memory collection in 3002 ms • 621.5 MB used of 639.0 MB
10:17:28.246	gc	_isolateEntry • low memory collection in 2373 ms • 580.7 MB used of 591.4 MB
10:17:28.305	gc	main • low memory collection in 3205 ms • 44.9 MB used of 54.3 MB
10:17:28.305	gc	main • low memory collection in 3206 ms • 44.9 MB used of 54.3 MB
10:17:28.339	gc	_isolateEntry • low memory collection in 2465 ms • 23.9 MB used of 27.3 MB
10:17:28.339	gc	_isolateEntry • low memory collection in 2466 ms • 23.9 MB used of 27.3 MB
10:17:28.368	gc	main • low memory collection in 3268 ms • 44.9 MB used of 54.3 MB
10:17:28.368	gc	main • low memory collection in 3268 ms • 44.9 MB used of 54.3 MB
10:17:28.424	gc	_isolateEntry • low memory collection in 2550 ms • 23.9 MB used of 27.3 MB
10:17:28.425	gc	_isolateEntry • low memory collection in 2550 ms • 23.9 MB used of 27.3 MB
10:17:28.435	gc	_isolateEntry • low memory collection in 2561 ms • 23.9 MB used of 27.3 MB
10:17:28.436	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285048436
10:17:28.437	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:28.438	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:28.438	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285048438
10:17:28.448	gc	main • low memory collection in 3348 ms • 44.9 MB used of 54.3 MB
10:17:28.448	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285048448
10:17:28.448	stdout	flutter: -------------------------- startImageStream: processed camera data in 349ms --------------------------
10:17:28.450	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285048450
10:17:28.499	gc	_isolateEntry • finalize collection in 2577 ms • 25.6 MB used of 28.5 MB
10:17:28.502	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285048502
10:17:28.504	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:28.505	gc	main • finalize collection in 3354 ms • 52.4 MB used of 61.4 MB
10:17:28.506	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:28.510	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285048509
10:17:28.511	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285048511
10:17:28.511	stdout	flutter: -------------------------- startImageStream: processed camera data in 61ms --------------------------
10:17:28.532	stdout	flutter: ZZZZZ startImageStream: sending image data to processing isolate. 1579285048532
10:17:28.534	stdout	flutter: ZZZZZ runImageProcessing entered. 1579285048534
10:17:28.535	stdout	flutter: ZZZZZ runImageProcessing: concatenate plane data: 1ms
10:17:28.536	stdout	flutter: ZZZZZ processImage: ffi call: 0ms
10:17:28.536	stdout	flutter: ZZZZZ runImageProcessing: leaving. 1579285048536
10:17:28.537	stdout	flutter: ZZZZZ startImageStream: returned from isolate. 1579285048537
10:17:28.537	stdout	flutter: -------------------------- startImageStream: processed camera data in 5ms --------------------------

.
.
.
```

flutter analyze:
```
Analyzing flutter_gc_camera_example...                                  

   info • Prefer double quotes where they won't require escape sequences • image_processing_plugin/lib/allocation.dart:5:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • image_processing_plugin/lib/allocation.dart:6:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • lib/camera_preview_test.dart:1:8 • prefer_double_quotes
   info • The value of the local variable 'result' isn't used • lib/image_worker.dart:108:10 • unused_local_variable
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:8:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:9:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:11:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:14:15 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:19:22 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:20:22 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:27:22 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:28:22 • prefer_double_quotes
```

flutter doctor -v:
```
[✓] Flutter (Channel master, v1.14.2-pre.34, on Mac OS X 10.15.1 19B88, locale en-CA)
    • Flutter version 1.14.2-pre.34 at <path>
    • Framework revision bf1d822198 (13 hours ago), 2020-01-16 19:38:01 -0800
    • Engine revision 85a8ac4255
    • Dart version 2.8.0 (build 2.8.0-dev.3.0 f1df196ddf)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at <path>
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.2.1, Build version 11B500
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] VS Code (version 1.40.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (4 available)
    • Lenovo TB 8504F • android-arm64 • Android 8.1.0 (API 27)
    • SM T290         • android-arm64 • Android 9 (API 28)
    • iPad 7th gen    • ios           • iOS 13.3
    • iPad 5th gen    • ios           • iOS 11.4.1
```

**Comment by mkustermann** [comment]
Just briefly looking at the code, one thing sticks out in [image_worker.dart:concatenatePlanes](https://github.com/martin-labanic/camera_preview_ffi_image_processing/blob/master/lib/image_worker.dart#L84):
```dart
  static Uint8List concatenatePlanes(List<Plane> planes) {
    final WriteBuffer allBytes = WriteBuffer();
    planes.forEach((Plane plane) => allBytes.putUint8List(plane.bytes));
    return allBytes.done().buffer.asUint8List();
  }
```
This uses an `WriteBuffer` abstraction, which uses `Uint8Buffer` from `package:typed_data` (which is implemented in a very generic way, see [Uint8Buffer.addAll](https://github.com/dart-lang/typed_data/blob/master/lib/typed_buffers.dart#L96))

It will probably do unnecessary copies, because this `WriteBuffer` (and the `Uint8Buffer` inside it) grow the buffer possibly multiple times. Which creates more garbage for the GC to collect.

A better way would be to do the concatenation manually, by first calculating the sum of the plane buffers, allocating a `Uint8List` and filling it once.

Though there are still copies: When the main isolate sends the `CameraImage` to the helper isolates all the bytes will be copied once. Then we copy them once again to create a concatenated bytes array. Then we copy them again to C memory.

At least the last copy we can avoid by concatenating to C memory directly, something like this:
```dart
  static ffi.Pointer<ffi.Uint8> concatenatePlanes(List<Plane> planes) {
    int totalBytes = 0;
    for (int i = 0; i < planes.length; ++i) {
      totalBytes += planes[i].bytes.length;
    }
    final pointer = allocate<ffi.Uint8>(count: totalBytes);
    final bytes = pointer.asTypedList(totalBytes);
    int byteOffset = 0;
    for (int i = 0; i < planes.length; ++i) {
      final length = planes[i].length
      bytes.setRange(byteOffset, byteOffset += length, planes[i].bytes);
    }
    return pointer;
  }
```

That will do less copying and thereby also create less garbage for the GC to collect.

**Comment by martin-labanic** [comment]
@mkustermann Good catch, I've pushed up a commit to the sample project with that change and updated the output.

**Comment by mkustermann** [comment]
You could avoid another allocation by doing the concatenation to C memory in the main isolate and just send the address of that pointer to the other isolate. But from your log it seems the concatenation takes longer than processing the image, is that right?

**Comment by rmacnak-google** [comment]
There are two VM issues here:

 - The growth policy is performing badly on this workload, always using parallel marking (long safepoint) instead of concurrent marking (short safepoint).
- As part of work towards sharing code between isolates, isolates in AOT that are related by Isolate.spawn run in an "isolate group" where they have to wait for each other's safepoints.

**Comment by martin-labanic** [comment]
@mkustermann I'll definitely look in to shifting that check to C++ code. For the example project the concatenation is the longest running part at the moment. When I have the third party library in there it takes a small fraction of the overall time. The best case scenario on the 5th gen ipad that I see with the library included tends to be between 30 - 32ms. ~5ms of it is in the dart code, and 25ms in the C++. The worst case, where the time to process a frame is over 100ms, I would say the C++ code tends to take between 90 - 96% of the time.

@rmacnak-google It sounds like changing an isolates "isolate group" would help resolve this. Looking at the isolate documentation, is `spawnUri` the approach I should take with this? https://api.flutter.dev/flutter/dart-isolate/Isolate/spawnUri.html

**Comment by martin-labanic** [comment]
@rmacnak-google I tried the changes on the master channel which include the commits you made. Overall its a substantial performance boost with android averaging close to the performance of native and iOS has improved by ~50 - 55%. The performance spikes appear to be hitting higher though.

Statistics from three separate tests run on a 5th gen iPad using the master channel (flutter v1.13.10-pre.67, dart 2.8.0 (build 2.8.0-dev.1.0 fe666ce592)).
Note: this is the example project with some additional diagnostics and a 3rd party image processing library used in the ffi call.
```
// 1
# of processed camera frames: 4000
# of discarded camera frames: 7704
min time for frame processing (including message passing): 32ms
max time for frame processing (including message passing): 1972ms
avg time for frame processing (including message passing): 74.2335ms
total time: 296934ms

// 2
# of processed camera frames: 4000
# of discarded camera frames: 8569
min time for frame processing (including message passing): 33ms
max time for frame processing (including message passing): 2550ms
avg time for frame processing (including message passing): 81.11475ms
total time: 324459ms

// 3
# of processed camera frames: 4000
# of discarded camera frames: 7257
min time for frame processing (including message passing): 33ms
max time for frame processing (including message passing): 618ms
avg time for frame processing (including message passing): 69.57425ms
total time: 278297ms
```

**Comment by kf6gpe** [comment]
@mkustermann @rmacnak-google Just wondering if you had any feedback for @martin-labanic's last comment. It seems odd that the fix is misbehaving on just one platform like this.

**Comment by dnfield** [comment]
Related: https://github.com/dart-lang/sdk/issues/42078

**Comment by dnfield** [comment]
@martin-labanic - when I try to run your demo on master, I get an abort due to thread violations. Looks like your platform view thread is trying to access something it shouldn't. Does this run correctly for you on master today?

**Comment by martin-labanic** [comment]
@dnfield I tried the demo on android (android 9) and ios (11.4.1) and I did not encounter any aborts.

flutter doctor -v:
```
[✓] Flutter (Channel master, 1.19.0-2.0.pre.214, on Mac OS X 10.15.2 19C57, locale en-CA)
    • Flutter version 1.19.0-2.0.pre.214 at <path>
    • Framework revision 1d395c5e18 (25 hours ago), 2020-05-31 07:41:50 -0700
    • Engine revision 923687f0e7
    • Dart version 2.9.0 (build 2.9.0-11.0.dev 6489a0c68d)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at <path>
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[!] VS Code (version 1.40.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
```

flutter analyze:
```
Analyzing flutter_gc_camera_example...                                  

   info • Prefer double quotes where they won't require escape sequences • image_processing_plugin/lib/allocation.dart:5:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • image_processing_plugin/lib/allocation.dart:6:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • lib/camera_preview_test.dart:1:8 • prefer_double_quotes
   info • The value of the local variable 'result' isn't used • lib/image_worker.dart:111:10 • unused_local_variable
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:8:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:9:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:11:8 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:14:15 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:19:22 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:20:22 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:27:22 • prefer_double_quotes
   info • Prefer double quotes where they won't require escape sequences • test/widget_test.dart:28:22 • prefer_double_quotes
```

**Comment by chinmaygarde** [comment]
Triage: @rmacnak-google Are there any updates? Just trying to figure out what to do with this issue.

**Comment by rmacnak-google** [comment]
I believe the second half was fixed by dart-lang/sdk@3451c348dbed2d80632f8141d6ab69651442eabc.

**Comment by martin-labanic** [comment]
@rmacnak-google Is this change in the newest flutter stable release?

**Comment by lumarama** [comment]
Still see the memory leak every time I run compute - my app memory footprint grows and never goes down after GC:
I've reduced my code to just calling empty function - and the issue is still reproducible. No memory leak if I don't call compute.
```
      await compute(_encodeTempJpeg, null);
      
      Future<void> _encodeTempJpeg(_JpegEncodeParams params) async {
      }
```

[√] Flutter (Channel stable, 1.22.5, on Microsoft Windows [Version 10.0.18363.1256], locale en-US)
    • Flutter version 1.22.5 at C:\sdk\flutter
    • Framework revision 7891006299 (6 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

**Comment by Aperico-com** [comment]
This morning I updated camera from 0.6.4+5 to 0.7.0 and I get loads of GC when I use controller.startImageStream((){}) and the CameraPreview stutters and has small freezes. I did not change my code at all, I only updated to 0.7.0. I even rolled back to 0.6.4+5 to double check and the garbage collection and freezing is gone. It gets a little better if I use jpg stream instead of yuv420 but still GC freeze. I have written a plugin for ML kit and do not see a feasible way to create a minimal test case at the moment, it seems very similar to the original posters though. I have timed the calls to native and they go from 800ms to 1300ms on average but the calls are done async and it is the GC that cause the freezes.
